enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,543 @@
+- Introducing an OLM owned CRD forces operators to integrate with OLM at a much deeper level than writing to a resource that they already manage. Operators would be required to detect if it is being managed by OLM and begin writing to the API. By writing to an owned CR, the Operator does not require OLM specific code and can be run independently of OLM.
+
+Allowing operators to write to the status of any CR creates a potential scaling issue if enough CRs are on cluster. OLM will avoid these performance issues by introducing the Status Probe Operator (SPO) which will reconcile the `Probe` resource. If an operator chooses to write to their own CRs, SPO will create a pod that will keep the status of the `Probe` resource synced with the conditions in the CRs.
+
+>Note: Unfortunately, OLM will not support an operator communicating its status via cluster-scoped CRs. If an operator is scoped to a namespace but writes conditions to a cluster-scoped CR, OLM would be unable to identify which operator is communicating its status if multiple installs of the operator were present on cluster.
+
+### Opting into OLM Supported Conditions
+
+Operators may opt-in to "OLM Supported Conditions" by adding annotations to their owned CRD(s), where the key is the defined by OLM and the value is a boolean expression over condition types applied to the CR at `.status.conditions`. The "OLM Supported Condition" annotation key will adhere to the following format: `operators.coreos.com/condition.<ConditionName>`.
+
+Let's consider the case where an operator is upgradeable when the migrating condition on a CR is false, the "OLM Supported Condition" annotation should be `operators.coreos.com/condition.Upgradeable: "!Migrating"`. Alternatively, the operator could  indicate that it is upgradeable when both the `ReadyToGo` and `FinishedMigrating` conditions are true by adding the `operators.coreos.com/condition.Upgradeable: "FinishedMigrating && ReadyToGo"` annotation.
+
+### Example Workflow
+
+Let's examine the complete workflow for an operator that owns the `Foo` crd where:
+
+- The operator should not be upgraded when a `Foo` CR has the `Migrating` condition set to true.
+- OLM should surface any CR condition whose `Type` matches `BadConnectivity` or `UnhealthyDatabase` in the `Probe` resource.
+
+In this case, the operator's CRD should include the following annotations:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata: {
+  name: foo
+  annotations: {
+    "operators.coreos.com/condition.Upgradeable" : "!Migrating" # The operator is upgradeable when the CR is not being migrated.
+    "operators.coreos.com/condition.Important" : "BadConnectivity || UnhealthyDatabase" # The Important condition should only include the `OR` boolean operator.
+  }
+}
+```
+
+When this annotation is present on a CRD, OLM will append the following `Probe` CR to the `installPlan`:
```

Comment:
What happens at update time if my operator shifts to having a probe resource in the bundle, are we able to handle that case?

During upgrade do we sync the CRD and we would lose any overrides? 

What is the value of making these annotations on the CRD's and not adding the probe resource to opt-in?

Reply:
> What happens at update time if my operator shifts to having a probe resource in the bundle, are we able to handle that case?

`Probes` will be owned by CSVs. During an upgrade, the old Probe will be acquired by the new CSV and updated based on the spec provided in the bundle. 

>  During upgrade do we sync the CRD and we would lose any overrides?

Once the probe is updated during an upgrade, the specOverride field would be set to the new values defined in the bundle. If the new probe had the specOverride field set to "crdAnnotations", then yes the sync would change overrides. If the new probe does not have the specOverride field set, then the crdAnnotations would not influence behavior.

> What is the value of making these annotations on the CRD's and not adding the probe resource to opt-in?

I might need to tap in @exdx and @njhale for this one, as the conversation we had was quite some time ago. There is no reason a user could not opt-in to this feature by including a probe resource in their bundle. The annotation approach without the Probe resource was a means to minimize the efforts of operator authors to integrate with the new OLM feature.

Reply:
Similarly, if the annotations define the probe when it's not provided in a bundle, can't we just generate probes dynamically at any time (not just in the InstallPlan)?